google-http-client/src/main/java/com/google/api/client/http/HttpBackOffIOExceptionHandler.java

```diff
@@ -97,6 +97,8 @@ public boolean handleIOException(HttpRequest request, boolean supportsRetry) thr
     try {
       return BackOffUtils.next(sleeper, backOff);
     } catch (InterruptedException exception) {
+      // Catching InterruptedException clears the thread interrupted bit.
```

Comment:
I could well be wrong here. I usually am when it comes to Thread arcana, but reading that article and the docs, this only seems necessary when you call `interrupted()`, not when you simply catch InterruptedException. That is, this code does not clear the interrupted bit. Am I missing something?

Reply:
This may well be the right thing to do.

Unless this code has control, authority, and context or institutes a policy so that eventually it can make the current interrupted thread die (or decide not to terminate the thread as its normal operation)–which I don't think is the case as a library, or unless this code re-throws `InterruptedException`, or unless it is really correct that this code is designed to consume thread interruption and exit by returning false, I think it should restore the interrupted bit, so that other parts of the code (someone higher in the call stack) can be made aware that this thread is interrupted and act accordingly.

https://stackoverflow.com/questions/2523721/why-do-interruptedexceptions-clear-a-threads-interrupted-status
https://stackoverflow.com/questions/4906799/why-invoke-thread-currentthread-interrupt-in-a-catch-interruptexception-block

But I think the comment `// Catching ... clears the thread interrupted bit.` is unnecessary.

Reply:
And I actually think the comment "Catching InterruptedException clears the thread interrupted bit" is incorrect and misleading. It's not the act of catching it. Rather, I believe it is just that the interrupted bit is already cleared (and should by convention) when `InterruptedException` is thrown. From https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html,

> By convention, any method that exits by throwing an InterruptedException clears interrupt status when it does so.

Reply:
It would help if this had a test demonstrating both that the bug is real and that this patch fixes it.

Reply:
Thanks for the quick response and thoughtful discussion.

Sorry about the delay. I havent had spare cycles to work on this recently...
I will write a test, but note that it is difficult to predict threading behavior so the test may be flaky at times, depending on cpu speed and available resources.

Reply:
Testing thread issues is extremely tricky. A flaky test can be worse than no test. This might need an integration test that doesn't run as part of the normal suite or some other magic.